### ISSUE TITLE
[ERROR hir_ty::mir] Only tuple or closure has tuple or closure field

### ISSUE THREAD

**Issue by Veykril** [rust-lang/rust-analyzer#15090]
While working on r-a itself I tend to see this spammed, I haven't figured out which part of the codebase causes this though

**Comment by M4n5ter** [comment]
I am currently working on a small demo project using Diesel and Axum. However, I am encountering some issues when trying to integrate Diesel. Specifically, rust-analyzer seems unable to resolve some Diesel names and  lose track of the types.

In addition, I am also seeing the following error: "[ERROR hir_ty::mir] Only tuple has tuple field".

**Comment by janosimas** [comment]
I have the same issue with `quick_error`

**Comment by LucaCappelletti94** [comment]
I see the same issue working with Diesel and Actix, maybe it has something to do with all of those macros from Diesel?

**Comment by adwhit** [comment]
I can reproduce with the following

```
impl Bork {
    fn f((x, y): T) {}
}
```

Note that `Bork` need not exist for the error to appear. It is specifically the pattern matching an unknown type within the signature that seems to cause the error.

**Comment by adwhit** [comment]
Investigated a little bit. Here is a way to make a similar error:

```
enum Foo {
    Bar { x: i32 },
}

fn f(Foo::Bar { x }: T) {}
```

Error: `internal error: entered unreachable code: Only adt has field, found {error}\n`.

And this:

```
fn f(&x: T) {}
```
Error: `internal error: entered unreachable code: Overloaded deref on type {unknown} is not a projection\n`

In general, the error-handling in this block is off: https://github.com/rust-lang/rust-analyzer/blob/2b7b44bf27a7635eded9aa53678acb2c4e25063e/crates/hir-ty/src/mir.rs#L154-L227. It assumes type errors cannot happen, but evidently they can.

**Comment by Veykril** [comment]
We might wanna set a flag for `InferenceResult` whether errors happened and then just error out in mir lowering if that is the case

**Comment by adwhit** [comment]
Here is another test case that, unlike the others, is actually valid code:

```
struct Foo<T> {
    l: T,
    r: T,
}

trait Trait {
    type State;
    fn f(_: Self::State) -> Self;
}

impl<T> Trait for Foo<T> {
    type State = (T, T);

    fn f(state: Self::State) -> Self {
        Self {
            l: state.0,
            r: state.1,
        }
    }
}
```

**Comment by EndilWayfare** [comment]
> It is specifically the pattern matching an unknown type within the signature that seems to cause the error.
> 
> Backtrace
> Edit: In fact it can be reduced to just
> 
> ```
> fn f((x, y): T) {}
> ```
Aww man, that's _totally_ why this slams `axum` and `diesel`, lmao. Extractors and SQL rows. I have a _lot_ of tuple destructuring in my data layer, so at least now I know _why_ my inference has ground to a crawl.

I've been meaning to get up to speed on `rust-analyzer` internals and dip my toes into contributing. How crazy of a first issue would this be to jump into, do you think?

**Comment by Veykril** [comment]
Hmm, fwiw this should only affect MIR building for the function so at worst you'll miss out on some diagnostics there.

The simple fix would be to skip mir building if we encounter any sort of type mismatches as said here https://github.com/rust-lang/rust-analyzer/issues/15090#issuecomment-1985638015 which shouldn't be too difficult. It's basically adding a bool field to `InferenceResult`, and check that in `lower_to_mir`.

**Comment by Dushistov** [comment]
I have the same issue.

I edit 

```
fn f(arg: (usize, bool)) {}
```

remove `arg` and add `(index, fla`
and rust-analyzer crash on

```
fn f((index,fla: (usize, bool)) {}
```

panic backtrace:

```
thread 'Worker' panicked at crates/hir-ty/src/mir.rs:185:21:
internal error: entered unreachable code: Only tuple has tuple field
stack backtrace:
   0: _rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::unreachable_display
   3: hir_ty::mir::ProjectionElem<V,T>::projected_ty
   4: hir_ty::mir::borrowck::moved_out_of_ref::{{closure}}
   5: hir_ty::mir::borrowck::moved_out_of_ref
   6: hir_ty::mir::borrowck::borrowck_query::{{closure}}
   7: hir_ty::mir::borrowck::all_mir_bodies
   8: hir_ty::mir::borrowck::borrowck_query
   9: salsa::Cycle::catch
  10: salsa::derived::slot::Slot<Q,MP>::execute
  11: salsa::derived::slot::Slot<Q,MP>::read
  12: <salsa::derived::DerivedStorage<Q,MP> as salsa::plumbing::QueryStorageOps<Q>>::fetch
  13: <DB as hir_ty::db::HirDatabase>::borrowck::__shim
  14: <DB as hir_ty::db::HirDatabase>::borrowck
  15: hir::DefWithBody::diagnostics
  16: hir::ModuleDef::diagnostics
  17: hir::Module::diagnostics
  18: ide_diagnostics::diagnostics
  19: salsa::Cancelled::catch
  20: ide::Analysis::diagnostics
  21: core::ops::function::impls::<impl core::ops::function::FnMut<A> for &mut F>::call_mut
  22: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
  23: rust_analyzer::diagnostics::fetch_native_diagnostics
  24: core::ops::function::FnOnce::call_once{{vtable.shim}}
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Comment by ChayimFriedman2** [comment]
@hkaiser25 This doesn't crash rust-analyzer unless you build from source, and there is a feature flag when building to not crash it.

**Comment by hkaiser25** [comment]
> [@hkaiser25](https://github.com/hkaiser25) This doesn't crash rust-analyzer unless you build from source, and there is a feature flag when building to not crash it.

Which feature flag? Will adding this flag to the “cargo check” command that rust analyzer runs in the background fix the crashes for me?

**Comment by ChayimFriedman2** [comment]
@hkaiser25 No, you need to set the feature flag when building rust-analyzer itself. If you install it from `xtask` it sets the feature, so you need to edit this line:

https://github.com/rust-lang/rust-analyzer/blob/c5882732e6e6e09ac75cddd13545e95860be1c42/xtask/src/install.rs#L140

And remove `--features force-always-assert`.

**Comment by lnicola** [comment]
@hkaiser25 you should file a new issue.

**Comment by hkaiser25** [comment]
#19543

**Comment by ChayimFriedman2** [comment]
This could be fixed by #22101, so I'm closing this and will reopen if we'll find it still occurs.

**Comment by lnicola** [comment]
@ChayimFriedman2 think we should also close #19543?

**Comment by ChayimFriedman2** [comment]
@lnicola That one was about `cargo check` failing, not about the log.
